platform/o.n.swing.laf.flatlaf/src/org/netbeans/swing/laf/flatlaf/ui/FlatTabControlIcon.java

```diff
@@ -0,0 +1,270 @@
+    private static final int arc = UIManager.getInt("TabControlIcon.arc"); // NOI18N
+    private static final Color foreground = UIManager.getColor("TabControlIcon.foreground"); // NOI18N
+    private static final Color disabledForeground = UIManager.getColor("TabControlIcon.disabledForeground"); // NOI18N
+    private static final Color rolloverBackground = UIManager.getColor("TabControlIcon.rolloverBackground"); // NOI18N
+    private static final Color pressedBackground = UIManager.getColor("TabControlIcon.pressedBackground"); // NOI18N
+    private static final Color closeRolloverBackground = UIManager.getColor("TabControlIcon.close.rolloverBackground"); // NOI18N
+    private static final Color closeRolloverForeground = UIManager.getColor("TabControlIcon.close.rolloverForeground"); // NOI18N
+
+    private static final Map<Entry<Integer,Integer>,Icon> INSTANCES = populateInstances();
+    private final int buttonId;
+    private final int buttonState;
+    private final float userScaleFactor; // for Java 8 and Linux
+
+    private static void populateOne(
+            Map<Entry<Integer,Integer>,Icon> toMap, int buttonId, int buttonState)
+    {
+        toMap.put(new SimpleEntry<Integer,Integer>(buttonId, buttonState),
+                new FlatTabControlIcon(buttonId, buttonState));
+    }
+
+    private static Map<Entry<Integer,Integer>,Icon> populateInstances() {
+        // The string keys of these maps aren't currently used, but are useful for debugging.
+        Map<String, Integer> buttonIDs = new LinkedHashMap<String, Integer>();
+        // ViewTabDisplayerUI
+        buttonIDs.put("close", TabControlButton.ID_CLOSE_BUTTON);
+        // These don't seem to be in use anymore.
+        //buttonIDs.put("slide_right", TabControlButton.ID_SLIDE_RIGHT_BUTTON);
+        //buttonIDs.put("slide_left", TabControlButton.ID_SLIDE_LEFT_BUTTON);
+        //buttonIDs.put("slide_down", TabControlButton.ID_SLIDE_DOWN_BUTTON);
+        buttonIDs.put("pin", TabControlButton.ID_PIN_BUTTON);
+        buttonIDs.put("restore_group", TabControlButton.ID_RESTORE_GROUP_BUTTON);
+        buttonIDs.put("slide_group", TabControlButton.ID_SLIDE_GROUP_BUTTON);
+        // EditorTabDisplayerUI
+        buttonIDs.put("scroll_left", TabControlButton.ID_SCROLL_LEFT_BUTTON);
+        buttonIDs.put("scroll_right", TabControlButton.ID_SCROLL_RIGHT_BUTTON);
+        buttonIDs.put("drop_down", TabControlButton.ID_DROP_DOWN_BUTTON);
+        buttonIDs.put("maximize", TabControlButton.ID_MAXIMIZE_BUTTON);
+        buttonIDs.put("restore", TabControlButton.ID_RESTORE_BUTTON);
+        Map<String, Integer> buttonStates = new LinkedHashMap<String, Integer>();
+        buttonStates.put("default", TabControlButton.STATE_DEFAULT);
+        buttonStates.put("pressed", TabControlButton.STATE_PRESSED);
+        buttonStates.put("disabled", TabControlButton.STATE_DISABLED);
+        buttonStates.put("rollover", TabControlButton.STATE_ROLLOVER);
```

Comment:
Should add `// NOI18N`s

Reply:
Just curious - this comment clutters the codebase - shouldn't it only be used at places where translation is expected?

Reply:
I think that we should add it if translation is not expected, but I could be wrong.

Reply:
The interesting part would be: why should we add this throughout the codebase. From my perspective it is distracting, causes friction and has a low return factor. I can see the use case when a class is translated, then it helps to keep untranslateable strings. But why do the work before it is clear, that translation is expected?

Reply:
I'm not sure... Maybe, oracle's people(@jlahoda @sdedic ) know something?

Reply:
I think if you use the Internationalization tooling in NetBeans across multiple files it will skip strings found on line with //NOI18N. So it may indeed be useful.

Reply:
@svenreimers I think so, too. Thank you.